Improvement/s3 utils 119 modify count items for storage metrics

XinLi0207 opened the pull request:
this PR implement a `CountItemsMongoClient` class which extends `MongoClientInterface` from arsenal
override `getObjectMDStats()` function  and some utils functions to output
```
{
        "objects" : 4,
        "versions" : 6,
        "buckets" : 3,
        "bucketList" : [ 
            {
                "name" : "bucket2",
                "location" : "us-east-1",
                "isVersioned" : true,
                "ownerCanonicalId" : "2fbfcc0125094acda54f0cabb5d1e66ae78aaf67f6060e8ac5b6abc9ada01d81",
                "ingestion" : false
            }, 
            {
                "name" : "bucket1",
                "location" : "us-east-1",
                "isVersioned" : true,
                "ownerCanonicalId" : "2fbfcc0125094acda54f0cabb5d1e66ae78aaf67f6060e8ac5b6abc9ada01d81",
                "ingestion" : false
            }
        ],
        "dataManaged" : {
            "total" : {
                "curr" : 392,
                "prev" : 222
            },
            "byLocation" : {
                "us-east-1" : {
                    "curr" : 392,
                    "prev" : 222
                }
            }
        },
        "stalled" : 0,
        "dataMetrics" : {
            "bucket" : {
                "bucket1" : {
                    "usedCapacity" : {
                        "current" : 376,
                        "nonCurrent" : 0
                    },
                    "objectCount" : {
                        "current" : 2,
                        "nonCurrent" : 0,
                        "deleteMarker" : 0
                    }
                },
                "bucket2" : {
                    "usedCapacity" : {
                        "current" : 6,
                        "nonCurrent" : 0
                    },
                    "objectCount" : {
                        "current" : 1,
                        "nonCurrent" : 0,
                        "deleteMarker" : 0
                    }
                },
            },
            "location" : {
                "us-east-1" : {
                    "usedCapacity" : {
                        "current" : 392,
                        "nonCurrent" : 222
                    },
                    "objectCount" : {
                        "current" : 4,
                        "nonCurrent" : 6,
                        "deleteMarker" : 1
                    }
                }
            },
            "account" : {
                "account2" : {
                    "usedCapacity" : {
                        "current" : 392,
                        "nonCurrent" : 222
                    },
                    "objectCount" : {
                        "current" : 4,
                        "nonCurrent" : 6,
                        "deleteMarker" : 1
                    }
                }
            }
        }
    }
}
```
added 3 metric levels: bucket/location/account and collect their usedCapacity and objectCount respectively

Commit:
improvement: S3UTILS-119 modify countItems for storage consump metrics

bert-e commented:
# Hello xinliscality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/s3utils/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/S3UTILS-119-modify-countItems-for-storage-metrics` and the
destination branch `development/1.13`.

Please resolve the conflict on **the feature branch** (`improvement/S3UTILS-119-modify-countItems-for-storage-metrics`).

```sh
git fetch && \
git checkout origin/improvement/S3UTILS-119-modify-countItems-for-storage-metrics && \
git merge origin/development/1.13
```

Resolve merge conflicts and commit

```sh
git push origin HEAD:improvement/S3UTILS-119-modify-countItems-for-storage-metrics
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

anurag4DSB commented on CountItems/CountManager.js:
```diff
@@ -61,6 +66,27 @@ class CountManager {
                 }
             });
         }
+        if (results.dataMetrics
+            && results.dataMetrics.bucket
+            && results.dataMetrics.location
+            && results.dataMetrics.account) {
+            Object.keys(results.dataMetrics).forEach(metricLevel => {
+                Object.keys(results.dataMetrics[metricLevel]).forEach(resourceName => {
+                    if (this.store.dataMetrics[metricLevel][resourceName]) {
+                        Object.keys(this.store.dataMetrics[metricLevel][resourceName]).forEach(subMetric1 => {
+                            Object.keys(this.store.dataMetrics[metricLevel][resourceName][subMetric1]).forEach(subMetric2 => {
+                                this.store.dataMetrics[metricLevel][resourceName][subMetric1][subMetric2]
+                                    += results.dataMetrics[metricLevel][resourceName][subMetric1][subMetric2];
+                            });
+                        });
+                    } else {
+                        this.store.dataMetrics[metricLevel][resourceName] = results.dataMetrics[metricLevel][resourceName];
+                    }
+                });
+            });
+        } else {
+            this.store.dataMetrics = results.dataMetrics;
+        }
```
It would be great to have a comment here, to explain this logic and why it is needed.

anurag4DSB replied:
This commend might apply here as well https://github.com/scality/Arsenal/pull/2055/files#r1054615483

XinLi0207 commented:
hi @alexanderchan-scality, as per your comment [here](https://github.com/scality/Arsenal/pull/2055#issuecomment-1363309467), I implemented a new `CountItemsMongoClient` which extends `MongoClientInterface` from arsenal and overrides `getObjectMDStats()` and some other functions, I agree with what you said, it's indeed better to be closer to CountItems logic and should not be a common mongo client function.
I have also moved all the tests I have done in this [PR](https://github.com/scality/Arsenal/pull/2055) here as well, so now there is no change need to be done at arsneal side, I will close that PR if this PR looks good to you.

williamlardier commented on CountItems/CountManager.js:
```diff
@@ -61,6 +70,25 @@ class CountManager {
             });
         }
+        if (results.dataMetrics
+            && results.dataMetrics.bucket
+            && results.dataMetrics.location
+            && results.dataMetrics.account) {
```
This will avoid updating the store dataMetrics if either of these 3 fields are not in the object, is this something expected? (-> or if I put in differently, can `dataMtrics` not have these three fields?)

XinLi0207 replied:
yeah it's kind of expected, because there won't be a case that there are metrics in buckets but not in location nor account, all objects must be linked to a bucket, an account and one or multiple locations. so if there is only bucket metrics, we assume the dataMetrics is broken

williamlardier reviewed (commented):
Mostly completes my first review on arsenal before the code migration. LGTM otherwise!

My main "concern" is on the `CountItemMongoClient` that duplicates some code with Arsenal while extending the class: this may be affected by changes in the Arsenal class after a bump in s3util if we are lacking tests on some parts, as we also rely on class methods.

XinLi0207 commented on tests/utils/mongoUtils.js:
```diff
@@ -0,0 +1,33 @@
+        count: 1,
+        dbName: 'metadata',
+        storageEngine: 'ephemeralForTest',
+    },
```
seems `mongodb-memory-server` doesn't support specifying the mongo version we are currently using, 4.0.1

nicolas2bert commented on utils/S3UtilsMongoClient.js:
```diff
@@ -0,0 +1,246 @@
+        if (!isTransient
+            || entry.value.replicationInfo.status !== 'COMPLETED') {
+            // only count it in current dataStore if object is not in transient or replication not completed
+            if (results.location[entry.value.dataStoreName]) {
```
When does this condition return true?
Is `_processEntryData` not re-initialized for every call?

XinLi0207 replied:
yeah, this condition can never be true, `results.location[entry.value.dataStoreName]` should simply be overwrited

nicolas2bert commented on utils/S3UtilsMongoClient.js:
```diff
@@ -0,0 +1,246 @@
+        entry.value.replicationInfo.backends.forEach(rep => {
+            // count it in the replication destination location if replication compeleted
+            if (rep.status === 'COMPLETED') {
+                if (results.location[rep.site]) {
```
Same for this one:
The location source and destination should be different and each destination location should be different.
Worst case, if the location happens to be duplicated, we should overwrite the size.

XinLi0207 replied:
yeah you are right, we should simply overwrite the size for every destination location, there is no need to accumulate it in one bucket.

nicolas2bert commented on utils/S3UtilsMongoClient.js:
```diff
@@ -0,0 +1,246 @@
+                        && this._isReplicationEntryStalled(res, cmpDate)) {
+                        stalledCount++;
+                    }
+                } else if (!!res.value.versionId && !res.value.isNull) {
```
Why do we check for isNull here?

From what I remember, "null versions" reference the objects located in an unversioned bucket.

XinLi0207 replied:
the previous `countItmes` service has an issue when bucket is suspended, because of [this logic ](https://github.com/scality/Arsenal/blob/development/8.1/lib/storage/metadata/mongoclient/MongoClientInterface.js#L2077) when location's nonCurrent size is got by versioned size minus master size(it works in versioned bucket, because master data are duplicated), but the object uploaded after the bucket is suspended will only have master data in DB, so this record is not duplicated so should not be subtracted

that's why I take out of `isNull` data from master data and treat it as non versioned data

nicolas2bert replied:
Would you mind adding a comment here?

XinLi0207 replied:
yeah, no problem
Thanks for reviewing!

nicolas2bert commented on utils/S3UtilsMongoClient.js:
```diff
@@ -0,0 +1,246 @@
+                        deleteMarkerCount,
+                    } = res[metricLevel][resourceName];
+
+                    dataMetrics[metricLevel][resourceName].usedCapacity.current += nullData + masterData;
```
`nullData` is referencing the size of the objects in an unversioned bucket.
`masterData` is referencing the size of the master versions in a versioned/suspended bucket.

For a given bucket state one or the other shoule be 0. Is it why we add them here?

XinLi0207 replied:
yeah, `nullData` represent current data in nonversioned bucket/ object uploaded after bucket suspended.
`masterData` represents current data in versioned bucket/object uploaded before bucket suspended
if one has value, the other is 0

nicolas2bert commented on utils/S3UtilsMongoClient.js:
```diff
@@ -0,0 +1,246 @@
+                        dataMetrics[metricLevel][resourceName].usedCapacity.nonCurrent += versionData - masterData;
+                        dataMetrics[metricLevel][resourceName].usedCapacity.nonCurrent = Math.max(dataMetrics[metricLevel][resourceName].usedCapacity.nonCurrent, 0);
+                        dataMetrics[metricLevel][resourceName].objectCount.nonCurrent
+                            += versionCount - masterCount - deleteMarkerCount;
```
Why do we need to subtract deleteMarkers from versions here? deleteMarker(s) are not considered as versions?

XinLi0207 replied:
we have 3 metrics for `usedCapacity`: current, nonCurrent, deleteMarker
I feel like it's more logical to have them separated, we can add them up if needed

williamlardier reviewed (approved):
I thought I already approved, but no :smile:

XinLi0207 commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/1.13`



The following branches will **NOT** be impacted:

* `development/1.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.13`



The following branches have **NOT** changed:

* `development/1.4`



Please check the status of the associated issue S3UTILS-119.

Goodbye xinliscality.
